go/vt/vtgate/engine/concatenate.go

```diff
@@ -408,6 +424,9 @@ func (c *Concatenate) coerceAndVisitResults(
 				}
 			}
 		}
+		if r.Fields != nil {
+			r.Fields = fields
+		}
 		err := callback(r)
```

Comment:
nit: this can be set only for the first result set.

Reply:
this can happen outside the for loop.

Reply:
Fixed